flax/serialization.py

```diff
@@ -246,6 +246,74 @@ def _msgpack_ext_unpack(code, data):
   return msgpack.ExtType(code, data)
 
 
+# Chunking array leaves
+
+# msgpack has a hard limit of 2**31 - 1 bytes per object leaf.  To circumvent
+# this limit for giant arrays (e.g. embedding tables), we traverse the tree
+# and break up arrays near the limit into flattened array chunks.
+
+# True limit is 2**31 - 1, but leave a margin for encoding padding.
+MAX_CHUNK_SIZE = 2**30
+
+
+_tuple_to_dict = lambda tpl: dict([(str(x), y) for x, y in enumerate(tpl)])
+_dict_to_tuple = lambda dct: tuple(dct[str(i)] for i in range(len(dct)))
+
+
+def _chunk(arr):
+  """Convert array to a canonical dictionary of chunked arrays."""
+  if isinstance(arr, jax.xla.DeviceArray):
+    arr = np.array(arr)
```

Comment:
I'm always a bit worried about the serialisation logic operating on jax arrays. Should we just move things to numpy at the very start of serialisation?

Reply:
sure, added a pass to do this first.